### ISSUE TITLE
Multi GPU gives a lot of:  blocking_queue.cpp:50] Waiting for data messages , training is slower than single GPU

### ISSUE THREAD

**Issue by siddharthm83** [BVLC/caffe#3264]
My system
1. 2x Titan X
2. CUDA 7.5
3. CuDNN v3
With 1 x TitanX 20 iterations for training on Imagenet-1000 with Caffenet takes about 6.5ms. 
`I1029 15:17:08.761509 20493 solver.cpp:236] Iteration 40, loss = 6.93335
I1029 15:17:08.761553 20493 solver.cpp:252]     Train net output #0: loss = 6.93335 (* 1 = 6.93335 loss)
I1029 15:17:08.761564 20493 sgd_solver.cpp:106] Iteration 40, lr = 0.01
I1029 15:17:14.897974 20493 solver.cpp:236] Iteration 60, loss = 6.95387
I1029 15:17:14.898021 20493 solver.cpp:252]     Train net output #0: loss = 6.95387 (* 1 = 6.95387 loss)
I1029 15:17:14.898032 20493 sgd_solver.cpp:106] Iteration 60, lr = 0.01`

When I use the -gpu all flag when training a network, time for 20 iterations is anywhere from 8ms to 10ms.  

`I1030 11:39:50.682731 24445 solver.cpp:511] Iteration 160 (3.41461/s), lr = 0.01
I1030 11:39:50.683522 24456 blocking_queue.cpp:50] Waiting for data
I1030 11:39:50.683884 24453 blocking_queue.cpp:50] Waiting for data
I1030 11:39:50.685421 24456 blocking_queue.cpp:50] Waiting for data
I1030 11:39:50.686043 24453 blocking_queue.cpp:50] Waiting for data
I1030 11:39:50.687865 24456 blocking_queue.cpp:50] Waiting for data
I1030 11:39:50.688470 24453 blocking_queue.cpp:50] Waiting for data
I1030 11:39:50.690054 24456 blocking_queue.cpp:50] Waiting for data
.........
I1030 11:39:58.427631 24453 blocking_queue.cpp:50] Waiting for data
I1030 11:39:58.428709 24456 blocking_queue.cpp:50] Waiting for data
I1030 11:39:58.429364 24453 blocking_queue.cpp:50] Waiting for data
I1030 11:39:58.430766 24456 blocking_queue.cpp:50] Waiting for data
I1030 11:39:58.431671 24445 solver.cpp:217] Iteration 180, loss = 6.90891`

My GPU utilization is not at 100% for either of the GPUs. I have 1 4TB HDD where my data is stored. Attached is the benchmark. Is the below enough or should I go to a RAID5 system for multi-GPU? 
![screenshot from 2015-10-30 14 03 23](https://cloud.githubusercontent.com/assets/5252864/10858019/08e46786-7f0f-11e5-84e5-9d4564ca4829.png)

**Comment by cypof** [comment]
What data_source do you use?

**Comment by thatguymike** [comment]
Is your LMDB compressed or raw images?  For compressed images, your disk should be fast enough but your CPU might not be fast enough to decode.  For uncompressed images, you need ~100MB/s per TitanX to keep things fed for imagenet+caffenet.  It is odd that it is so much slower...

**Comment by siddharthm83** [comment]
@cypof : yes, all default caffenet settings. I created lmdb using the caffe script which I am assuming is not compressing images.

@thatguymike Thanks. From my HDD benchmark, it appears that my average read is 141.1MB/sec. So I guess that is good for 1 Titan X but not sufficient for 2 Titan X based on your comment?

**Comment by aaalgo** [comment]
I've done some measurement regarding image IO: http://www.wdong.org/wordpress/blog/2015/11/07/image-storage-for-deep-learning-raw-or-jpeg/ .

In your case, I believe the best solution is to switch to an SSD, which will boost your throughput to about 500MB/sec.  An alternative as thatguymike suggested is to store JPEG images into LMDB, and you'll save 4x in size.  If you link against libturbo jpeg, a single core can decode at about 1350 images/second.  (141MB/s raw image translates to 752/s.)  However, disk I/O and image decoding have to be properly pipelined to get the full throughput.  I'm not sure how this is handled in Caffe.  I'm working on a image storage called PicPoc to replace LMDB (and a caffe fork), but that's not matured yet.

**Comment by thatguymike** [comment]
Disk IO and decode are not fully overlapped in the current code.  We would need to thread a bit deeper than we do now.  Going to larger prefetch depth might also help, but currently that will trigger a bug.  

If you disable the test phase (set iterations to 0 in your solver), do you still get the warning messages?
